dosbox-staging/dosbox-staging
# Implement DOS file/directory attributes support

# Description

This is a next generation version of the patch originally prepared by @NicknineTheEagle https://github.com/dosbox-staging/dosbox-staging/pull/2693

First short explanation: Archive bit in MS-DOS means _file is newly created or modified, and archivizer software should make a backup copy during next run and clean the bit afterwards_ (at least this was the idea, in reality most of the files in MS-DOS system have Archive bit set).

This PR implements the following:
- all DOS file/directory attributes are now supported (not just Read-Only, as before)
- newly created or modified files get the Archive bit set (on both local drive and FAT images), I have tried to duplicate MS-DOS behavior as closely as possible
- on Windows host attributes are stored normally in the filesystem
- on Linux / macOS hosts Read-Only is denoted by clearing write permissions (as before), moreover all the attributes are stored in _xattr_ extended attributes, in a format compatible with WINE and DOSEmu 2 (they both use Samba 3 format); lack of extended attributes is considered as Archive bit set and Hidden / System not set (as before this implementation)

Limitations:
- overlay drives (`-t overlay` option in `MOUNT` command) are partially supported; attributes can be retrieved (always), but can be only set for files in the overlay - before anyone complains: current implementation does not even allow to set file in the overlay as Read-Only (request is mostly ignored), so the situation is actually improved now :)
- interoperability with WINE does not cover Archive bit; WINE always assumes it is set for the files and does not allow to change it
- interoperability with DOSEmu 2 wasn't tested, as I am unable to set up the emulator (crashes for me instantly after startup)

## Related issues
Implements https://github.com/dosbox-staging/dosbox-staging/issues/2793

# Manual testing

1. Just play with `ATTRIB` command or other tools to change attributes; all newly created or modified files (for example using `EDIT.COM` from MS-DOS) should get the Archive attribute set. Tested on Linux and Windows (yes, despite I'm a Linux user, I have launched it, waited till it installs 6 months of updates, and did some tests!). I was unable to test under macOS, I don't have access to Mac.

2. To test interoperability with WINE, the easiest way is to symlink some directory from DOSBox local drive to WINE virtual drive, execute `wine cmd` command from console, and play with `ATTRIB` commands of both emulators; note that WINE ignores Archive attribute (always has it set for files) and does not allow to change it.

3. I wasn't sure how MS-DOS precisely handles Archive attribute (does it set it after you open file for writing? after you write to the file? when you are closing the modified file?), so I have written a test program (attached below). To try it:
- rename file to `ATEST.C`
- install OpenWatcom inside DOSBox
- execute `OWSETENV.BAT` from the compiler directory
- compile the test program using `WCL.EXE -l=dos ATEST.C` command
- run the executable; for convenience, it shows reference values - attributes in hex format, retrieved while running the program under real MS-DOS 6.22

[ATEST.C.TXT](https://github.com/dosbox-staging/dosbox-staging/files/12688978/ATEST.C.TXT)

# Checklist

I have:
